ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
 	identd->filter_list = ecalloc(elistsz, sizeof(mbfl_identify_filter *));
 
 	/* create filters */
-	i = 0;
-	num = 0;
-	while (i < elistsz) {
-		filter = mbfl_identify_filter_new2(elist[i]);
-		if (filter != NULL) {
-			identd->filter_list[num] = filter;
-			num++;
-		}
-		i++;
+	for (int i = 0; i < elistsz; i++) {
+		identd->filter_list[i] = mbfl_identify_filter_new2(elist[i]);
 	}
-	identd->filter_list_size = num;
-
-	/* set strict flag */
+	identd->filter_list_size = elistsz;
 	identd->strict = strict;
 
 	return identd;
 }
 
-
-void
-mbfl_encoding_detector_delete(mbfl_encoding_detector *identd)
+void mbfl_encoding_detector_delete(mbfl_encoding_detector *identd)
 {
-	int i;
-
-	if (identd != NULL) {
-		if (identd->filter_list != NULL) {
-			i = identd->filter_list_size;
-			while (i > 0) {
-				i--;
-				mbfl_identify_filter_delete(identd->filter_list[i]);
-			}
-			efree((void *)identd->filter_list);
-		}
-		efree((void *)identd);
+	for (int i = identd->filter_list_size - 1; i >= 0; i--) {
+		mbfl_identify_filter_delete(identd->filter_list[i]);
 	}
+	efree((void*)identd->filter_list);
+	efree((void*)identd);
 }
 
-int
-mbfl_encoding_detector_feed(mbfl_encoding_detector *identd, mbfl_string *string)
+int mbfl_encoding_detector_feed(mbfl_encoding_detector *identd, mbfl_string *string)
 {
-	int res = 0;
-	/* feed data */
-	if (identd != NULL && string != NULL && string->val != NULL) {
+	if (string->len > 0) {
 		int num = identd->filter_list_size;
-		size_t n = string->len;
-		unsigned char *p = string->val;
+		unsigned char *p = string->val, *e = p + string->len;
 		int bad = 0;
-		while (n > 0) {
-			int i;
-			for (i = 0; i < num; i++) {
+
+		/* We have an array of filter functions corresponding to different encodings
+		 * Iterate over the string, pushing each byte into each and every filter function
+		 *
+		 * If the string is not encoding XYZ, then likely at some point we will hit a
+		 * byte which is not valid in encoding XYZ, and the the 'flag' for that encoding
+		 * will be set.
+		 *
+		 * Our hope is that at some point, all the candidate encodings will be flagged
+		 * except one, which will presumably be the correct one. */
+		while (p < e) {
+			for (int i = 0; i < num; i++) {
 				mbfl_identify_filter *filter = identd->filter_list[i];
 				if (!filter->flag) {
-					(*filter->filter_function)(*p, filter);
+					(filter->filter_function)(*p, filter);
 					if (filter->flag) {
 						bad++;
 					}
 				}
 			}
-			if ((num - 1) <= bad) {
-				res = 1;
-				break;
+			/* If all candidate encodings have failed, or there is only one remaining
+			 * which has not failed, we can return.
+			 *
+			 * The exception is in strict mode; because when we discover that all
+			 * other possible encodings have failed, we might still be in the middle
+			 * of a multi-byte character for the 'correct' encoding, and stopping
+			 * right here would cause `mbfl_encoding_detector_judge` (below) to
+			 * wrongly reject it. */
```

Comment:
I think the better explanation here would be that we still need to check whether the remainder of the string is valid (presumably that's what strict mode is for).

Reply:
Amended comment to include your thought. However, it is also true that strict mode rejects candidate encodings if the input string ends in the middle of a multi-byte character when interpreted in that encoding... and non-strict mode does not.

Reply:
Yes, that is part of what constitutes a valid multi-byte string :) It's just not the main thing this is about.